theano/gpuarray/dnn.py

```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+                            context_name=context_name)()
+        dgrid = GpuArrayType(dtype=self.dtype,
+                             broadcastable=img.type.ndim * (False,),
+                             context_name=context_name)()
+
+        inputs = [img, theta, grid, grid_dims, dy, desc, alpha, beta]
+        outputs = [dimg, dgrid]
+
+        return Apply(self, inputs, outputs)
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1, 1], [1, 1], [1, 1], [1, 1], [1, 1], [0, 0], [1, 1], [1, 1]]
+
+
+class GpuDnnTransformerGradT(DnnBase):
+    """
+    Gradients of the affine transformation generated by the localisation network
+    """
+    __props__ = ('dtype',)
+    _cop_num_inputs = 2
+    _cop_num_outputs = 1
+    _f16_ok = True
+
+    def __init__(self, dtype=theano.config.floatX):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gt.c"], "APPLY_SPECIFIC(dnn_sptf_gt)")
+        self.dtype = dtype
+
+    def make_node(self, dgrid, desc):
+        context_name = infer_context_name(dgrid)
+        dtheta = GpuArrayType(dtype=self.dtype,
+                              broadcastable=(dgrid.type.ndim - 1) * (False,),
+                              context_name=context_name)()
+        inputs = [dgrid, desc]
+        outputs = [dtheta]
+
+        return Apply(self, inputs, outputs)
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1], [0]]
+
+
+def dnn_spatialtf(img, theta, scale_width=1, scale_height=1, alpha=None, beta=None,
+                  dtype=theano.config.floatX):
+    """
+    GPU spatial transformer using cuDNN from NVIDIA.
+
+    Parameters
+    ----------
+    img : tensor
+        Images to which the transformations will be applied.
+    theta : tensor
+        Affine transformation tensor containing one affine transformation
+        matrix per image. ``theta`` is usually generated by the localization
+        network.
+    scale_height: float
+        A float specifying the scaling factor for the height of the output
+        image. A value of 1 will keep the original height of the input. Values
+        larger than 1 will upsample the input. Values below 1 will downsample
+        the input.
+    scale_width: float
+        A float specifying the scaling factor for the width of the output
+        image. A value of 1 will keep the original width of the input. Values
+        larger than 1 will upsample the input. Values below 1 will downsample
+        the input.
+
+    Returns
+    -------
+    out : tensor
+        Transformed images with width and height properly scaled.
+
+    Notes
+    -----
+    Currently, cuDNN only supports 2D transformations with 2x3 affine
+    transformation matrices.
+
+    Also, the only grid sampler method available is the bilinear interpolation.
+    """
+
+    # inp is a 4D tensor with shape: (num_inputs, num_channels, height, width)
+    assert img.ndim == 4
```

Comment:
At this point you are not sure that img has a `ndim` field. You should apply `as_gpuarray_variable()` before.